proxy/src/main/java/com/velocitypowered/proxy/text/translation/TranslationManager.java

```diff
@@ -0,0 +1,596 @@
+        } else {
+          throw new UnsupportedFormatException("Invalid format: " + content.substring(start, end));
+        }
+      }
+
+      if (appender != null) {
+        if (end != content.length()) {
+          appender.append(content, end, content.length());
+        }
+
+        appender.appendRemaining();
+        TextComponent.Builder builder = appender.builder();
+        builder.style(component.style());
+
+        List<Component> translatedChildren = translateComponentsIfNeeded(
+            registry, children, formattingContext);
+        builder.append(translatedChildren != null ? translatedChildren : children);
+
+        if (hoverEvent != null) {
+          HoverEvent translatedHoverEvent = translateHoverEventIfNeeded(
+              registry, hoverEvent, formattingContext);
+          builder.hoverEvent(translatedHoverEvent != null ? translatedHoverEvent : hoverEvent);
+        }
+
+        return builder.build();
+      }
+    }
+
+    List<Component> translatedChildren = translateComponentsIfNeeded(
+        registry, children, formattingContext);
+
+    HoverEvent translatedHoverEvent = hoverEvent != null ? translateHoverEventIfNeeded(
+        registry, hoverEvent, formattingContext) : null;
+
+    if (component instanceof TranslatableComponent) {
+      TranslatableComponent translatable = (TranslatableComponent) component;
+
+      String key = translatable.key();
+      Translation translation = registry.get(key);
+      if (translation != null) {
+        Component formatted;
+        if (translation instanceof FixedTranslation) {
+          FixedTranslation fixedTranslation = (FixedTranslation) translation;
+          formatted = fixedTranslation.component;
+        } else {
+          List<Component> arguments = translatable.args();
+          FormattingContext context = new FormattingContext(arguments);
+          formatted = translateComponent(registry, translation.component, context);
+        }
+        if (children.isEmpty()) {
+          if (!component.hasStyling()) {
+            return formatted;
+          }
+        } else {
+          children = translatedChildren != null ? translatedChildren :
+              new ArrayList<>(children);
+          children.addAll(0, formatted.children());
+          formatted = formatted.children(children);
+        }
+        if (translatedHoverEvent != null) {
+          formatted = formatted.hoverEvent(translatedHoverEvent);
+        }
+        return formatted.style(mergeStyle(component.style(), formatted.style()));
+      }
+    }
+
+    if (translatedChildren != null || translatedHoverEvent != null) {
+      if (translatedChildren != null) {
+        component = component.children(translatedChildren);
+      }
+      if (translatedHoverEvent != null) {
+        component = component.hoverEvent(translatedHoverEvent);
+      }
+      return component;
+    }
+
+    return null;
+  }
+
+  // TODO: Replace with text library method once added
```

Comment:
With the latest version of the text library, this can be replaced with `first.merge(second)`